File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    term = {}
    for t in terms:
        p = t.split(" ")
        term[p[0]] = int(p[1])
    
    for idx, privacy in enumerate(privacies):
        p = privacy.split(" ")
        if isDeprecated(today, calculate_date(p[0], term[p[1]])):
            answer.append(idx+1)
            
    return answer


def isDeprecated(today, end_date):
    d = today.split(".")
    for i in range(3):
        d[i] = int(d[i])
    if (d[0] > end_date[0]):
        return True
    if (d[0] == end_date[0]):
        if d[1] > end_date[1]:
            return True
        if d[1] == end_date[1]:
            if d[2] > end_date[2]:
                return True
    return False

def calculate_date(day, limit):
    d = day.split(".")
    for i in range(3):
        d[i] = int(d[i])
    
    d[2] -= 1
    if d[2] == 0:
        d[1] -= 1
        d[2] = 28
    d[1] += limit
    if d[1] > 12:
        d[0] += (d[1]-1) // 12
        d[1] = 12 if d[1] % 12 == 0 else d[1] % 12
    return d

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;
import java.time.LocalDate;

class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        ArrayList<Integer> res = new ArrayList<>();
        LocalDate now = toDate(today);

        Map<String, Integer> termMap = new HashMap<>();
        for (String term : terms) {
            String[] split = term.split(" ");
            termMap.put(split[0], Integer.parseInt(split[1]));
        }

        for (int i=0;i<privacies.length;i++) {
            String[] info = privacies[i].split(" ");
            LocalDate exp = toDate(info[0]).plusMonths(termMap.get(info[1]));
            if (now.isAfter(exp) || now.isEqual(exp)) {
                res.add(i+1);
            }
        }
        return res.stream().mapToInt(Integer::intValue).toArray();
    }

    public LocalDate toDate(String date) {
        String[] ymd = date.split("[.]");
        int y = Integer.parseInt(ymd[0]);
        int m = Integer.parseInt(ymd[1]);
        int d = Integer.parseInt(ymd[2]);
        return LocalDate.of(y, m, d);
    }
}